src/realm/index_string.cpp

```diff
@@ -1187,11 +1188,34 @@ void StringIndex::find_all_fulltext(std::vector<ObjKey>& result, StringData valu
 
     auto tokenizer = Tokenizer::get_instance();
     tokenizer->reset({value.data(), value.size()});
-    while (tokenizer->next()) {
-        auto token = tokenizer->get_token();
-        FindRes res1 = find_all_no_copy(StringData{token.data(), token.size()}, res);
+    auto token_info = tokenizer->get_token_info();
+    struct Token {
+        std::string_view token;
+        bool exclude;
+    };
+    std::list<Token> tokens;
+    auto exclude_start = tokens.begin();
+    // Order the tokens so that we handle words to match first
+    for (auto& info : token_info) {
```

Comment:
Was there any particular reason to not integrate this into Tokenizer class itself? Looks to like natural things to do if we want to use it also for search terms parsing.

Reply:
I am not sure what you want to build into the Tokenizer.

Reply:
Add maybe some TokenizerWithExclusion on top of DefaultTokenizer. To me it doesn't seem to belong here, and i'd prefer whole parsing of search terms be done in tokenizer implementation rather that in index. It seems like we way tokenizer is used now is not fully efficient, so it's probably whole new separate refactoring is needed for this. Let just keep it in mind for now.